### PR TITLE
Erase flash pages progressively

### PR DESCRIPTION
**Pull request by nvlsianpu**
- Fixed ISR in serial_adapter to work with USB CDC ACM serial emulation. This commit prepares mcuboot for coming USB support patch.
- Added an option to erase flash pages progressively to avoid possible timeouts

This PR contains code which was separated form previous PR #318

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>
Signed-off-by: Emanuele Di Santo <andrzej.puzdrowski@nordicsemi.no>

### CODE REVIEW

**Comment by utzig** [comment]
It seems to me that by using `flash_area_sector_from_off` inside `CONFIG_BOOT_ERASE_PROGRESSIVELY` you make that code Zephyr only. This defeats the idea of target-OS code living only inside their respective directories; I'll give it some more thought soon. The other changes here look good.

**Comment by nvlsianpu** [comment]
@utzig I can call boot_read_sectors()->boot_initialize_area() for populate information about NVM structure, and then use this info instead of this dedicated flash_area_sector_from_off call.

File: boot/boot_serial/src/boot_serial.c [hunk 1]
```diff
@@ -323,7 +327,11 @@ bs_upload(char *buf, int len)
...
             goto out_invalid_data;
         }
+#ifndef CONFIG_BOOT_ERASE_PROGRESSIVELY
         rc = flash_area_erase(fap, 0, fap->fa_size);
+#else
+        rc = 0;
+#endif
         if (rc) {
             goto out_invalid_data;
         }
```
Comment (utzig) [inline boot/boot_serial/src/boot_serial.c]: This block above would be easier to read with just:

```
#ifndef CONFIG_BOOT_ERASE_PROGRESSIVELY
        rc = flash_area_erase(fap, 0, fap->fa_size);
        if (rc) {
            goto out_invalid_data;
        }
#endif
```

File: boot/boot_serial/src/boot_serial.c [hunk 2]
```diff
@@ -339,13 +347,57 @@ bs_upload(char *buf, int len)
...
+            /* get the last sector offset */
+            rc = flash_area_sector_from_off(boot_status_off(fap), &sector);
+            if (rc) {
+                BOOT_LOG_ERR("Unable to determine flash sector of"
+                             "the image trailer");
+                goto out;
+            }
+            /* assure that sector for image trailer was erased */
+            if (off_last < sector.fs_off) {
+                /* check  */
```
Comment (utzig) [inline boot/boot_serial/src/boot_serial.c]: check what?

File: boot/zephyr/flash_map_extended.c [hunk 3]
```diff
@@ -24,6 +23,16 @@
 #define FLASH_DEVICE_ID SOC_FLASH_0_ID
 #define FLASH_DEVICE_BASE CONFIG_FLASH_BASE_ADDRESS
 
+static struct device *flash_dev;
+
+struct device *flash_device_get_binding(char *dev_name)
+{
+    if (!flash_dev) {
+	    flash_dev = device_get_binding(dev_name);
```
Comment (utzig) [inline boot/zephyr/flash_map_extended.c]: tabs -> spaces

File: boot/boot_serial/src/boot_serial.c [hunk 4]
```diff
@@ -339,13 +347,57 @@ bs_upload(char *buf, int len)
             img_blen -= rem_bytes;
         }
     }
+
+#ifdef CONFIG_BOOT_ERASE_PROGRESSIVELY
+    static off_t off_last = -1;
+    struct flash_sector sector;
```
Comment (utzig) [inline boot/boot_serial/src/boot_serial.c]: To maintain the current style, please move those declarations to the functions' head...

**Comment by nvlsianpu** [comment]
@utzig updated

File: boot/boot_serial/src/boot_serial.c [hunk 5]
```diff
@@ -339,13 +349,55 @@ bs_upload(char *buf, int len)
...
+            if (rc) {
+                BOOT_LOG_ERR("Unable to determine flash sector of"
+                             "the image trailer");
+                goto out;
+            }
+            /* Assure that sector for image trailer was erased. */
+            /* Check whether it was erased during previous upload. */
+            if (off_last < sector.fs_off) {
+                
+                BOOT_LOG_INF("Erasing sector at offset 0x%x", sector.fs_off);
```
Comment (utzig) [inline boot/boot_serial/src/boot_serial.c]: This line has only spaces now...

Comment (nvlsianpu) [reply boot/boot_serial/src/boot_serial.c]: fixed

**Pull request merged**